### DISCUSSION TITLE
Proposal: New `multiselect` entity type

### DISCUSSION THREAD

**Discussion by raman325** [home-assistant/architecture#927] [Entity Models] [answered]
## Background

The `select` entity supports single select use cases but does not support multiselect use cases. A multiselect is useful in cases where e.g. days of the week, supported modes, or any other case where multiple items for the same parameter need to be selected.

## Proposal

Create a new `multiselect` entity type using the `multiselect` domain.

Properties:
- `options`: list of all available options
- `current_options`: list of all selected options of len >=0
- `supported_features`: `SUPPORT_EMPTY_SELECTION` would indicated whether len >=0 or len >= 1 is required.

State:
- JSON dump of `current_options`

Services:
 - `select_options`: Takes list of options as an input. Does error handling based on supported_features

## Alternatives considered

- Extend `select` platform: We could do this but the trend has been towards making simpler, building block type entities rather than overloading existing ones.

**Comment by elupus** [comment]
I would vote for extending current select entity with this. Feel like a reasonable extension.

One consideration, dont know if json for the state of such entity would be good since state is generally in the end a string.

**Reply by raman325** [reply]
right but what's the alternative? dump would be a string representation of the list so that it could easily be parsed

**Reply by raman325** [reply]
alternative is that we make it a comma separated string but store the list in an attribute to make access easier and not require deserialization each time

**Reply by ScottG489** [reply]
At least thinking of this from a technical perspective, I don't think this should be the same as the current select. The current select is like an enum and a multi-select is like an array/list (scalar vs vector).

Also from having worked with config flows, I know there's already a UI representation of this. Perhaps this should be a helper as well?

Otherwise, I know I've definitely found myself wanting some sort of collection entity type, so this would be really nice!

**Comment by TheFes** [comment]
What if the state would be the number of selected options, and the selected options would be in an attribute:
```yaml
name: Fruitbowl
state: 2
attributes:
  options:
    - Banana
    - Apple
    - Orange
    - Lemon
    - Durian
  selected:
    - Apple
    - Durian
```

**Reply by bdraco** [reply]
I think that lends a lot better to the new state translations model than some of the other proposals here.

Note sure on `state` or even if it needs to be anything more than `on`/`off` for if anything is selected at all or not

**Comment by frenck** [comment]
My grind with the proposal is:

> State:
> JSON dump of current_options

I don't think we should dump data structures in there... (but I also lack a better suggestion 😅)

**Comment by Petro31** [comment]
Ideas for state:

* Timestamp on last selection, allowing users to get list of selected items from `option` attribute.
   
   **Pros**:
   - Each state change would be unique and trigger.
   - We already have this behavior with Button entities

   **Cons**:
   - The state really has no meaning.

* Count of selected items, allowing users to get list of selected items from `option` attribute.

  **Pros**:
  - Each state change would be unique and trigger.
  - We already have this behavior with zones.

  **Cons**:
  -  Some state changes could be missed.  If a user somehow selects an item while simultaneously deselecting another item, the state will not change.

* A value that somehow tells the frontend how to translate the state into it's selections.  E.g.  AB means options 1 and 2 were selected, or a hex representation of each selection.  I.e.  55 items in the list would be a hex string representing binary of on/off for each selection, where the position in the binary representation is the index.  E.g. all 55 items are selected, state would be `7FFFFFFFFFFFFF`, if only the 55th item is selected... `40000000000000`.  We'd need to ensure that the value is not converted to an integer, and remain a hex string.

  **Pros**:
  - Each state change will cause a trigger.  No room for error here, would behave the same way as the timestamp version in regards to triggers.
  - Should easily work with enums when entity is provided by integration.
  - The translated state will show all options selected
  - The under the hood state will be templateable to get the actual selections
  - The under the hood state will always be less than 254 characters.

  **Cons**:
  - Will confuse people who look at the state under the hood.

Personally, I vote for the 3rd option as it will give everyone what they need and we could add a template filter/function that converts the value to a list of options (if needed).

**Reply by TheFes** [reply]
I must admit I'm not so familiar with hexadecimal binary representations, but I expect that with the 3rd option, adding an option, or changing the order of the options would also mean a state change, even if there was no actual change in the selected options.

That would not be the case for options 1 and 2.

**Reply by Petro31** [reply]
Yes, the idea is to capture every state change possible for triggers, while also maintaining the pro's bullet points.  We don't want a built in entity going over the 254 character limit yet we want to represent the selections in the main state.

I use this method to send signals in the software I write with multiselect when trying to minimize the transfer data.  It works well for serial coms and keeps the data low.  I usually use a series of bytes though instead of 1 big hex.

**Reply by TheFes** [reply]
Well, imagine a normal single choice select with options `[ 'foo', 'bar', 'foobar']`
If I select `foo` and then later add the option `whale` the state of this select entity will not change, so for a state trigger which is set to not trigger on attribute changes (with `for: null`) it will not trigger. Also when I would change the order of the options to eg `[ 'foobar', 'bar', 'foo' ]`

If the state of the multiselect entity would represent the datetime of the last change to the selected options, or the number of selected options, there would also be no state change if eg both `foo` and `bar` were selected and a `whale` was added as a 4th option (or the order was changed)

But if the hexadecimal representation is used, and again `foo` and `bar` are selected (and `foobar` not) the state would be `"6"`
However if then the option `whale` is added, the state will be recalculated to `"C"` Als when the order of the options changes for `[ 'foo', 'bar', 'foobar']` to `[ 'foobar', 'bar', 'foo' ]` the state would change to `"3"`
This would mean a state change, where the actual selected options did not change. 

Personally, I would expect the state to be based on the selected options, and not change if the number of options which can be selected, or the order of the options can change. As long as `foo` and `bar` are still the only two selected options, I would expect no state change. (and again, the state object will change of course, so in case of a state trigger without a setting like `for: null` there will be a trigger if the possible options are changed)

**Reply by TheFes** [reply]
On the other hand, with the number of selected options as the state, state changes can be missed.
I expect there will be a service call to select the options. Using that service call on the same example like above, someting like this will not change the number of selected options, and therefor no state change (if the number of selected options would be the state)
```yaml
service: multiselect.select_options
data:
  entity_id: multiselect.example
  options:
    - bar
    - foobar
  ```

**Reply by Petro31** [reply]
> But if the hexadecimal representation is used, and again `foo` and `bar` are selected (and `foobar` not) the state would be `"6"`
> However if then the option `whale` is added, the state will be recalculated to `"C"` Als when the order of the options changes for `[ 'foo', 'bar', 'foobar']` to `[ 'foobar', 'bar', 'foo' ]` the state would change to `"3"`
> This would mean a state change, where the actual selected options did not change.

If whale was added, it would be added as a higher int.  The state would not change.  It's not added at the beginning, the hex representation is opposite the list.  I.e.  your first item is the last bit.  Your last item is the first bit (going left to right).  Meaning the representation of the list would not change when items are added.  It would change if items are inserted.  I would expect a state change in that regard.  Regardless, if you add/remove items from the list, selections typically need to happen again anyways.  What happens when you remove a selection that was selected?  Normal behavior when removing/adding items is to clear your current selections.

**Reply by Petro31** [reply]
e.g.

Your list `[ a, b, c ]` where `a` and `b` are selected, would have a binary representation of `011` (i.e. cba, abc reversed) and a hex representation of `3`.

You decide to add `d` to the list.

Your list is now `[ a, b, c, d ]` where `a` and `b` are still selected.  Your binary representation of the is is now `0011` (dcba, abcd reversed) and a hex representation of `3`.

**Reply by TheFes** [reply]
Ah right, thanks!

**Reply by parautenbach** [reply]
I would definitely go for @Petro31's third suggestion. It's very common in application development (called feature flags) to control application features, rollouts, and such. In lower levels of development it would often be called bit masks and such (also used in networking e.g. your subnet mask). In other words, this is a very common and good use of this idea. In fact, the media player features I've worked with in HA when building my custom component does this too. 

In terms of the listed con, since states are anyways always strings, I'd hex encode it, and even consider prefixing it with `0x` which is commonly used to indicate a hex value. I'd do this rather than an integer representation. It is true the a string representation would be quite bloated compared to its true binary representation, but I think that's a small price to pay. 

Back to media players and feature flags: The attribute in that case is actually represented as an integer (example `supported_features: 450487`). In other words, it would be consistent to use an integer representation (and more compact), but personally I'd prefer the hex string.

**Reply by ScottG489** [reply]
I'm not sure I'm all for option 3. The value doesn't have any face value to the user, and I think it's too technical of a detail being exposed too high up in the stack. I also don't think we have a precedence for doing something like this for other states?

Also if we're comfortable exposing a technical detail like this at this level, why not just use JSON for the state?

I almost want to say option 1 would be the best of the 3 proposed here. I think we have some precedence for using timestamps for state and putting the selected values in the attributes seems to work well enough. That said, I'm still not really a fan of this and I feel like there has been some decisions recently that indicate we want to move away from attributes for this kind of information.

Otherwise, I don't really have an alternative to suggest. However, I do know I've wished I had ways to store more complicated data in HA (objects, collections, etc) without a great option. Perhaps there not being a clean solution here for how to represent the state indicates a deeper problem.

**Comment by TheFes** [comment]
It would be great if the options could have a different `label` and `value`
For example if you would want to have a multiselect to send a TTS message to speakers around your house
```yaml
options:
  - label: Livingroom
    value: media_player.livingroom
  - label: Bedroom
    value: media_player.this_is_where_the_magic_happens
  - label: Kids Room
    value: media_player.only_baby_shark_here
```

**Reply by ScottG489** [reply]
This is definitely a must. I believe this is possible with the existing multi-select that's available as part of the config flow (and maybe elsewhere).

**Comment by Phrede** [comment]
The perspective of a long time user... The long standing method across many platforms is using ctrl+ and shft+ to multi select items from a list.  It appears there may be some overthinking going on here.

This is sorely needed.  My example is that Amazon Alexa is greatly reducing their UI while increasing the problem areas.  There currently is no way to delete multiple devices.  Nor is there a global delete all that on could reinstall the needed ones.  This required me to delete 300+ devices ONE AT A TIME!   Amazon will not address this, of course, maybe HA will.

**Comment by starkillerOG** [comment]
@raman325 do you have any short term plans for implementing this?
I would love to use this for the Reolink integration, but if there are no short term plans, I will just implement a bunch of extra switch entities.

**Reply by raman325** [reply]
I can work up the selected proposal, but do we actually have consensus on this solution?

**Reply by starkillerOG** [reply]
I think the best solution is propsed by @TheFes here: https://github.com/home-assistant/architecture/discussions/927#discussioncomment-6505767

But I do not know when and how a final dicission is normally made on a arch discussion.

**Reply by gjohansson-ST** [reply]
So far I looks more like several ideas rather than a complete proper proposal somewhere

**Reply by AJediIAm** [reply]
A multi select is basically a group of booleans. Implementing it as such (group of switches) with a value of true if one of the options is selected could also be an option.

**Comment by mikey0000** [comment]
Same here this is perfect for selecting mowing zones!

**Comment by frenck** [comment]
As I spotted some activity on this proposal, I wanted to provide a heads up / status update atm.

We discussed this proposal in a recent core architectural meeting and decided not to act on it yet. So, we keep it in the current proposal state on purpose. The reasoning is that we have some more ideas and goals for our information architecture in the short term that might impact entity proposals (like this one).

../Frenck

**Reply by raman325** [reply]
ack thanks for the heads up. Let us know when this is ready to be worked again, I would be happy to work on the PoC once we decide on an approach

**Reply by starkillerOG** [reply]
Thanks for the update frenck!

**Reply by Daniel-dev22** [reply]
Any updates on this?

**Reply by egakupi** [reply]
Is there any updates?

**Comment by diestrohs** [comment]
In Lovelace under Automation → Time and Date → Trigger, there is already a graphical solution for multiselect that looks very charming. The only thing missing is the appropriate multiselect entity.
However, it would be beneficial to have an abbreviation of the texts for compact displays if they no longer fit into the display window.
![Multiselect-Dropdown](https://github.com/user-attachments/assets/84792146-7116-4166-ba07-f602d2fca17d)

**Comment by nprez83** [comment]
I ended up using multiple input_booleans, which I grouped into a single group, and then used fold-entiry-row to display them. Works and looks great in my opinion. If the group toggle is turned on, every input_boolean gets turned on, which is akin to disabling all my motion sensors (or the auto locks of all my doors). If one of the input booleans in the group gets turned on, the group toggle is turned on.

![image](https://github.com/user-attachments/assets/8f434baa-9cc6-4c2e-a786-71c6f76284f0)

![image](https://github.com/user-attachments/assets/fb3ecd27-6514-4ed3-ad5d-bc355156f45b)

**Comment by frenck** [comment, marked as answer]
I'm in the process of cleaning up our queues of architectural proposals a bit.

And my comment from 2024 still hold true: https://github.com/home-assistant/architecture/discussions/927#discussioncomment-10327739

This proposal is not going to move forward on a short term. Therefore, I'll go ahead and close it up for now. We can reconsider and/or reopen in the future.

../Frenck

**Comment by Domingato** [comment]
That unfortunate, while I understand choices have to be made.

I wanted to have a multi-selection feature on my dashboard. My last resort was trying to use this:
[https://github.com/mattieha/select-list-card](url)

and have 2 of those card with 2 select input entities working in tandem via an automation in order to emulate the functionality as seen in the image below. This is the kind of selection where selected elements go to the right and unselected go back to the left.

![image](https://github.com/user-attachments/assets/ac9eae6f-95eb-4300-a5cd-77cc777ca33f)

Very unfortunately this last select-list-card has not been maintained for many years and doesn't work properly today. It would be great if anybody would be able and willing to fix that.

**Comment by lboue** [comment]
New feature request here: [MultiSelect Entity #257](https://github.com/orgs/home-assistant/discussions/257)

**Reply by frenck** [reply]
That is not a discussion, but a feature request. There is quite the difference 😄

**Reply by lboue** [reply]
I've corrected my comment to avoid confusion as it's in the ‘discussion’ section.
